File: CBLClient/Apps/CBLTestServer-Android/app/src/main/java/com/couchbase/CouchbaseLiteServ/server/RequestHandler/PredictiveQueriesRequestHandler.java
package com.couchbase.CouchbaseLiteServ.server.RequestHandler;

/*
  Created by sridevi.saragadam on 2/26/19.
 */

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.couchbase.CouchbaseLiteServ.server.Args;
import com.couchbase.lite.CouchbaseLiteException;
import com.couchbase.lite.DataSource;
import com.couchbase.lite.Database;
import com.couchbase.lite.Dictionary;
import com.couchbase.lite.Expression;
import com.couchbase.lite.Function;
import com.couchbase.lite.PredictionFunction;
import com.couchbase.lite.PredictiveModel;
import com.couchbase.lite.Query;
import com.couchbase.lite.QueryBuilder;
import com.couchbase.lite.Result;
import com.couchbase.lite.ResultSet;
import com.couchbase.lite.SelectResult;


public class PredictiveQueriesRequestHandler {


    public EchoModel registerModel(Args args) {
        String modelName = args.get("model_name");
        EchoModel echoModel = new EchoModel(modelName);
        Database.prediction.registerModel(modelName, echoModel);
        return echoModel;
    }

    public void unRegisterModel(Args args) {
        String modelName = args.get("model_name");
        Database.prediction.unregisterModel(modelName);
    }

    public List<Object> getPredictionQueryResult(Args args) throws CouchbaseLiteException {
        EchoModel echoModel = args.get("model");
        Database database = args.get("database");
        Map<String, Object> dict = args.get("dictionary");
        Expression input = Expression.value(dict);
        PredictionFunction prediction = Function.prediction(echoModel.getName(), input);

        Query query = QueryBuilder
            .select(SelectResult.expression(prediction))
            .from(DataSource.database(database));

        List<Object> resultArray = new ArrayList<>();
        ResultSet rows = query.execute();
        for (Result row : rows) {
            resultArray.add(row.toMap());
        }
        return resultArray;
    }

    public String nonDictionary(Args args) {
        EchoModel echoModel = args.get("model");
        Database database = args.get("database");
        String dict = args.get("nonDictionary");
        Expression input = Expression.value(dict);
        PredictionFunction prediction = Function.prediction(echoModel.getName(), input);

        Query query = QueryBuilder
            .select(SelectResult.expression(prediction))
            .from(DataSource.database(database));

        List<Object> resultArray = new ArrayList<>();
        try {
            query.execute();
        }
        catch (Exception e) {
            return e.getLocalizedMessage();
        }
        return "success";
    }

    public List<Object> getEuclideanDistance(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");

        String key1 = args.get("key1");
        String key2 = args.get("key2");
        Expression distance = Function.euclideanDistance(Expression.property(key1), Expression.property(key2));

        Query query = QueryBuilder
            .select(SelectResult.expression(distance))
            .from(DataSource.database(database));

        List<Object> resultArray = new ArrayList<>();
        ResultSet rows = query.execute();
        for (Result row : rows) {
            resultArray.add(row.toMap());
        }
        return resultArray;
    }

    public int getNumberOfCalls(Args args) {
        EchoModel echoModel = args.get("model");
        return echoModel.numberOfCalls;
    }

    public List<Object> getSquaredEuclideanDistance(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");

        String key1 = args.get("key1");
        String key2 = args.get("key2");
        Expression distance = Function.squaredEuclideanDistance(Expression.property(key1), Expression.property(key2));

        Query query = QueryBuilder
            .select(SelectResult.expression(distance))
            .from(DataSource.database(database));

        List<Object> resultArray = new ArrayList<>();
        ResultSet rows = query.execute();
        for (Result row : rows) {
            resultArray.add(row.toMap());
        }
        return resultArray;
    }

    public List<Object> getCosineDistance(Args args) throws CouchbaseLiteException {
        Database database = args.get("database");

        String key1 = args.get("key1");
        String key2 = args.get("key2");
        Expression distance = Function.cosineDistance(Expression.property(key1), Expression.property(key2));

        Query query = QueryBuilder
            .select(SelectResult.expression(distance))
            .from(DataSource.database(database));

        List<Object> resultArray = new ArrayList<>();
        ResultSet rows = query.execute();
        for (Result row : rows) {
            resultArray.add(row.toMap());
        }
        return resultArray;
    }

    private static final class EchoModel implements PredictiveModel {
        private static final String TAG = "ECHO";
        public final String name;
        private int numberOfCalls = 0;

        EchoModel(String name) {
            android.util.Log.i(TAG, "Entered into echo model");
            this.name = name;
        }

        public String getName() {
            return name;
        }

        @Override
        public Dictionary predict(Dictionary input) {
            numberOfCalls++;
            return input;
        }
    }
}

File: CBLClient/Apps/CBLTestServer-iOS/CBLTestServer-iOS/Server/PredictiveQueriesRequestHandler.swift
//
//  PredictiveQueriesRequestHandler.swift
//  CBLTestServer-iOS
//
//  Created by Sridevi Saragadam on 2/20/19.
//  Copyright © 2017 Sridevi Saragadam. All rights reserved.
//

import Foundation
import CouchbaseLiteSwift


public class PredictiveQueriesRequestHandler {
    public static let VOID: String? = nil
    fileprivate var _pushPullReplListener:NSObjectProtocol?
    
    public func handleRequest(method: String, args: Args) throws -> Any? {
        switch method {
            
        case "predictiveQuery_registerModel":
            let modelName: String = args.get(name: "model_name")!
            let echoModel = EchoModel(name: modelName)
            Database.prediction.registerModel(echoModel, withName: modelName)
            return echoModel
        
        case "predictiveQuery_unRegisterModel":
            let modelName: String = args.get(name: "model_name")!
            Database.prediction.unregisterModel(withName: modelName)
        
        case "predictiveQuery_getPredictionQueryResult":
            let model: EchoModel = args.get(name: "model")!
            
            //let dict: [String: Any] = ["friend_one": Expression.property("friend_one")]
            
            let dict: [String: Any] = args.get(name: "dictionary")!
            let database: Database = args.get(name: "database")!
            let input = Expression.value(dict)
            let prediction = Function.prediction(model: model.name, input: input)
            let queryResult = QueryBuilder
                .select(SelectResult.expression(prediction))
                 .from(DataSource.database(database))
            
            var resultArray = [Any]()
            
            for row in try queryResult.execute() {
                resultArray.append(row.toDictionary())
            }
            
            return resultArray
            
        case "predictiveQuery_nonDictionary":
            let model: EchoModel = args.get(name: "model")!
            
            //let dict: [String: Any] = ["friend_one": Expression.property("friend_one")]
            
            let text: String = args.get(name: "dictionary")!
            let database: Database = args.get(name: "database")!
            let input = Expression.value(text)
            let prediction = Function.prediction(model: model.name, input: input)
            let queryResult = QueryBuilder
                .select(SelectResult.expression(prediction))
                .from(DataSource.database(database))
            // return try queryResult.execute()
            do{
                try queryResult.execute();
            }
            catch {
              return error.localizedDescription
            }
            
        case "predictiveQuery_getNumberOfCalls":
            let model: EchoModel = args.get(name: "model")!
            return model.numberOfCalls
            
        case "predictiveQuery_euclideanDistance":
            let database: Database = args.get(name: "database")!
            let key1: String = args.get(name: "key1")!
            let key2: String = args.get(name: "key2")!
            let key3_distance = Function.euclideanDistance(between: Expression.property(key1),
                                                      and: Expression.property(key2))
            let queryResult = QueryBuilder
                .select(SelectResult.expression(key3_distance))
                .from(DataSource.database(database))
            var resultArray = [Any]()
            
            for row in try queryResult.execute() {
                resultArray.append(row.toDictionary())
            }
            
            return resultArray
            
        case "predictiveQuery_squaredEuclideanDistance":
            let database: Database = args.get(name: "database")!
            let key1: String = args.get(name: "key1")!
            let key2: String = args.get(name: "key2")!
            let distance = Function.squaredEuclideanDistance(between: Expression.property(key1),
                                                           and: Expression.property(key2))
            let queryResult = QueryBuilder
                .select(SelectResult.expression(distance))
                .from(DataSource.database(database))
            var resultArray = [Any]()
            
            for row in try queryResult.execute() {
                resultArray.append(row.toDictionary())
            }
            
            return resultArray
            
        case "predictiveQuery_cosineDistance":
            let database: Database = args.get(name: "database")!
            let key1: String = args.get(name: "key1")!
            let key2: String = args.get(name: "key2")!
            let distance = Function.cosineDistance(between: Expression.property(key1),
                                                             and: Expression.property(key2))
            let queryResult = QueryBuilder
                .select(SelectResult.expression(distance))
                .from(DataSource.database(database))
            var resultArray = [Any]()
            
            for row in try queryResult.execute() {
                resultArray.append(row.toDictionary())
            }
            
            return resultArray
        
        default:
            throw RequestHandlerError.MethodNotFound(method)
        }
        return PredictiveQueriesRequestHandler.VOID
    }
}

class EchoModel: PredictiveModel {
    
    let name: String
    var numberOfCalls = 0
    
    init(name: String) {
        self.name = name
    }
    
    func predict(input: DictionaryObject) -> DictionaryObject? {
        numberOfCalls = numberOfCalls + 1
        return input;
    }
    
}
